core/src/Streamly/Internal/Control/Exception.hs

```diff
@@ -165,54 +186,110 @@ newtype AllocateIO = AllocateIO
 withAllocateIO :: (MonadIO m, MonadMask m) => (AllocateIO -> m a) -> m a
 withAllocateIO action = do
     -- Assuming 64-bit int counter will never overflow
-    ref <- liftIO $ newIORef (0 :: Int, Map.empty)
+    ref <- liftIO $ newIORef (0 :: Int, Map.empty, Map.empty)
     action (AllocateIO (allocator ref)) `MC.finally` aft ref
 
     where
 
-    -- The thread doing manual release
-    -- do not need to worry about concurrent execution.
-    aft ref = liftIO $ do
-        xs <- readIORef ref
-        sequence_ (snd xs)
-
-    allocator ref alloc free = do
-        (r, index) <- liftIO $ mask_ $ do
-            r <- alloc
-            idx <- atomicModifyIORef ref (\(i, mp) ->
-                ((i + 1, Map.insert i (void $ free r) mp), i))
-            return (r, idx)
-
-        let modify (i, mp) =
-                let res = Map.lookup index mp
-                 in ((i, Map.delete index mp), res)
-
-            release = do
-                f <- atomicModifyIORef ref modify
-                sequence_ f
+    -- XXX can we ensure via GC that the resources that we are freeing are all
```

Comment:
So we don't support parallel release yet.
PS: I'm reviewing the commits one by one. This commit may be obsolete.